### ISSUE TITLE
HLSL with wrong entry point generates code that SPIRV-Cross crashes on

### ISSUE THREAD

**Issue by hrydgard** [KhronosGroup/glslang#588]
The following HLSL:
```
float4 PixelShaderFunction(float4 input) : COLOR0   {
    return float4(0.0, 1.0, 0.0, 2.0);
}
```
with setEntryPoint("main") generates the following SPIR-V:
```
                              Capability Shader
               1:             ExtInstImport  "GLSL.std.450"
                              MemoryModel Logical GLSL450
                              EntryPoint Fragment 4  "main"
                              ExecutionMode 4 OriginUpperLeft
                              Name 4  "main"
                              Name 11  "PixelShaderFunction(vf4;"
                              Name 10  "input"
               2:             TypeVoid
               3:             TypeFunction 2
               6:             TypeFloat 32
               7:             TypeVector 6(float) 4
               8:             TypePointer Function 7(fvec4)
               9:             TypeFunction 7(fvec4) 8(ptr)
              13:    6(float) Constant 0
              14:    6(float) Constant 1065353216
              15:    6(float) Constant 1073741824
              16:    7(fvec4) ConstantComposite 13 14 13 15
         4(main):           2 Function None 3
               5:             Label
                              FunctionEnd
11(PixelShaderFunction(vf4;):    7(fvec4) Function None 9
       10(input):      8(ptr) FunctionParameter
              12:             Label
                              ReturnValue 16
                              FunctionEnd
```
This is invalid code, caused by the mismatched entry point name. Surely glslang should not generate invalid SPIR-V.

(Oh yeah, in addition, the return type should be SV_Target, too, not COLOR0, but I don't think that has anything to do with the issue).

**Comment by hrydgard** [comment]
Argh, never mind, the entry point name "main" didn't match "PixelShaderFunction". Doh.

Anyway, SPIRV-Cross should probably output some nicer error, but that's a different issue.

**Comment by hrydgard** [comment]
Reopening this, because glslang should clearly throw an error here instead of proceeding to generate invalid SPIR-V. See https://github.com/KhronosGroup/SPIRV-Cross/issues/68 .

**Comment by hrydgard** [comment]
Pared the issue report down to the interesting part.

**Comment by dneto0** [comment]
I've experienced this too.

**Comment by johnkslang** [comment]
> glslang should clearly throw an error here instead of proceeding to generate invalid SPIR-V

I'd say that glslang should either generate valid SPIR-V or throw an error. It was a minor and fixable bug to have it generate valid SPIR-V. So, I had it generate valid SPIR-V plus give a warning that the specified entry point was not found.  Does that work?

@durswd figured out one cause of invalid SPIR-V, that **OpReturn** was missing in the canonical entry point. That is fixed by fca826212cddbef03d69c4d7ab3a3efa5e2c81e7.

e50dc536ff76310bd24fa222397cca9d0b6acacf gives a warning when the specified entry point name is not found in the stage.

Is that sufficient to close this issue?

**Comment by hrydgard** [comment]
Having it warn and also generate valid SPIR-V is totally fine by me, closing.
